pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -679,6 +683,49 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
 	}
 }
 
+func (og *operationGenerator) markDeviceErrorState(volumeToMount VolumeToMount, devicePath, deviceMountPath string, mountError error, actualStateOfWorld ActualStateOfWorldMounterUpdater) {
+	if volumetypes.IsOperationFinishedError(mountError) &&
+		actualStateOfWorld.GetDeviceMountState(volumeToMount.VolumeName) == DeviceMountUncertain {
+		// Only devices which were uncertain can be marked as unmounted
+		markDeviceUnmountError := actualStateOfWorld.MarkDeviceAsUnmounted(volumeToMount.VolumeName)
```

Comment:
My thought is if you don't mark device as unmount when previously marked as uncertain. then you could simplify the logic. there is no need of TransientOperationError. MarkDeviceAsUnmounted will be triggered during volume tear down phase. 
What do you think?

Reply:
There are two reasons, I am not very sure about taking this path:
1. volumemanager has been designed to not track volumes that are not mounted (or for which mount failed with 100% certainity). Now we are saying - we will keep volumes which have failed to mount in "uncertain" state and hence they will require "teardown" and "unmountdevice" calls.  This introduces additional design change into volumemanager and I hope that - no plugin breaks (or fail unmount call) when unmount is called for failed mounts.  My intention has been to stick to code paths which are known to work.

2. The second reason is - what are we supposed to return when transient error does happen in mount calls? 
    * Mount is called and kubelet fails to fetch the secret: - If we return `uncertainProgressError` for them, then volume will be added to ASOW in uncertain state, even though mount call has not even been attempted.  It seems wrong to accept a volume in ASOW. This PR also made a change that, fetching secret call has been moved up and hence if there was an error fetching secret, call returns early without even creating volume directories.
    * If we return a regular error for transient errors - then that would cause volume directories to be removed. This is a problem if, previous error was "UncertainProgressError". We do not want a transient error to remove volume directories, in case previous mount operation is still in-progress.

Given above reasons - IMO having a transientError is not problematic. It makes it easier to reason about state of volumemanager and mount failures.

Reply:
Updated second case - since we moved secret fetching code to top of the function,  the defer is not executed if fetching secret fails.  But the main point IMO still stands.

Reply:
> volumemanager has been designed to not track volumes that are not mounted (or for which mount failed with 100% certainity). Now we are saying - we will keep volumes which have failed to mount in "uncertain" state and hence they will require "teardown" and "unmountdevice" calls. This introduces additional design change into volumemanager and I hope that - no plugin breaks (or fail unmount call) when unmount is called for failed mounts. My intention has been to stick to code paths which are known to work.

For this one, actually plugin has to deal with "teardown while previous mount still in progress or actually failed mount already" after adding this uncertain state, I think. The sequence of event would be like
1. volume setup call returns with timeout error, add it into state
2. pod is deleted, volume tear down is called (meantime the mount might still in process or might failed or might succeeded)

> Mount is called and kubelet fails to fetch the secret(on first attempt): - If we return uncertainProgressError for them, then volume will be added to ASOW in uncertain state, even though mount call has not even been attempted. It seems wrong to accept a volume in ASOW. This PR also made a change that, fetching secret call has been moved up and hence if there was an error fetching secret, call returns early without even creating volume directories.

I thought mount is called after fetching the secret?

Reply:
> I thought mount is called after fetching the secret?

I have been thinking about this and I think we still have a problem. Lets say - `MountDevice` is called and first attempt returns in `UncertainProgressError` which causes volume to be added to uncertain state. Second attempt however fails with a final error (not secret fetching error but lets say NodeStage itself failed) then because error was a final error - https://github.com/kubernetes/kubernetes/blob/6dbd521cfaf3b2abe69dce376b3b2a1cda5406e1/pkg/volume/csi/csi_attacher.go#L270 will remove the deviceMountPath from node. But volume is still in "Uncertain" state and must be cleaned up, but https://github.com/kubernetes/kubernetes/blob/6dbd521cfaf3b2abe69dce376b3b2a1cda5406e1/pkg/volume/csi/csi_attacher.go#L484 will throw an error because data directory from `deviceMountPath` is already removed..

Reply:
What do we actually gain by removing `TransientOperationFailure` from this PR?

Reply:
My thought was if both with and without transaction error can work correctly, I prefer without transaction for simplicity. Later we don't need to worry about later when other ppl work on this code, they need to understand this logic well and set correctly types of error.

But as you point it out, if mount fails, metadata will be cleaned. Then when reconciler calls unmount, it will fail to get metadata. The logic here can help remove volume from actual state cache, so unmount does not need to be triggered. 

So I think this is good. Just to make sure we add enough comment about different types of errors